### PR TITLE
Add collector, dispatcher and storage adapters

### PR DESCRIPTION
**Pull request by adrigzr**

### CODE REVIEW

**Review by eumendoza — DISMISSED** [review]

File: addon/storage-adapters/indexed-db.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	async isSupported() {
+		return this.db.open().then(() => true, () => false);
+	}
+
+	count(this: IndexedDb) {
+		return this.table.count();
+	}
+
+	push(this: IndexedDb, ...items: any[]) {
+		return this.db.transaction('rw', this.table, () => {
```
Comment (adrigzr) [inline addon/storage-adapters/indexed-db.ts]: Why `transaction` on a single `bulkAdd` operation?

```suggestion
		return this.table.bulkAdd(items);
```

File: addon/storage-adapters/indexed-db.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+			return;
+		}
+
+		const firstItem = await this.table.toCollection().first();
+
+		await this.db.transaction('rw', this.table, () => {
+			let _id = firstItem._id - 1;
+
+			items.reverse().forEach((item) => {
+				this.table.add(Object.assign({}, item, { _id }));
```
Comment (adrigzr) [inline addon/storage-adapters/indexed-db.ts]: This seems ok, but I would suggest to `reverse` and `map` items first, and then do `bulkAdd` without transaction.

File: addon/storage-adapters/indexed-db.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		}
+
+		return [];
+	}
+
+	async pop(this: IndexedDb, count?: number) {
+		const times = count || 1;
+		let result: any[] = [];
+
+		for (let i = 0; i < times; i++) {
```
Comment (adrigzr) [inline addon/storage-adapters/indexed-db.ts]: This should be in a `transaction`.

File: addon/storage-adapters/indexed-db.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		}
+		return result;
+	}
+
+	async shift(this: IndexedDb, count?: number) {
+		const times = count || 1;
+		let result: any[] = [];
+
+		for (let i = 0; i < times; i++) {
+			const item = await this.removeItem();
```
Comment (adrigzr) [inline addon/storage-adapters/indexed-db.ts]: This should be in a `transaction`.

File: addon/storage-adapters/local-storage.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+			throw new Error('LocalStorage storage adapter needs a key');
+		}
+
+		this.db = window.localStorage;
+	}
+
+	async isSupported() {
+		try {
+			window.localStorage.setItem('supported', '0');
+			window.localStorage.removeItem('supported');
```
Comment (adrigzr) [inline addon/storage-adapters/local-storage.ts]: This should be `this.db`.

File: addon/storage-adapters/local-storage.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	async count(this: LocalStorage) {
+		return this.getItems().length;
+	}
+
+	async push(this: LocalStorage, ...items: any[]) {
+		const storedItems = this.getItems();
+
+		items.forEach((item) => {storedItems.push(item);});
```
Comment (adrigzr) [inline addon/storage-adapters/local-storage.ts]: Split this line.

File: addon/storage-adapters/local-storage.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+		await this.setItems(storedItems);
+	}
+
+	async unshift(this: LocalStorage, ...items: any[]) {
+		const storedItems = this.getItems();
+
+		items.forEach((item) => {storedItems.unshift(item);});
```
Comment (adrigzr) [inline addon/storage-adapters/local-storage.ts]: Split this line.

File: addon/storage-adapters/memory.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+		items.forEach((item) => this.memory.unshift(item));
+	}
+
+	async pop(this: Memory, count?: number) {
+		const times = count || 1;
+
+		return this.memory.splice(-times);
+	}
+
+	async shift(this: Memory, count?: number) {
```
Comment (adrigzr) [inline addon/storage-adapters/memory.ts]: You can set a default on arguments:

```suggestion
	async shift(this: Memory, count: number = 1) {
```

File: addon/storage-adapters/indexed-db.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+		this.table.bulkAdd(items);
+	}
+
+	private async removeItem(this: IndexedDb, pop?: boolean) {
+		const collection = await this.table.toCollection();
+		const currentItem = pop ? await collection.last() : await collection.first();
+
+		if (isPresent(currentItem)) {
+			return this.table.delete(currentItem._id).then(() => {
```
Comment (adrigzr) [inline addon/storage-adapters/indexed-db.ts]: Use `async` here.

```suggestion
			await this.table.delete(currentItem._id);
```

**Review by eumendoza — DISMISSED** [review]

**Review by shokmaster — APPROVED** [review]

**Pull request merged**